staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/defaulting/surroundingobject.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package defaulting
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// FocusAccessorFunc encodes a JSON path through the given object, returning a leave,
+// with the bool being false if the path does not exist.
+type FocusAccessorFunc func(map[string]interface{}) (interface{}, bool, error)
+
+// SurroundingObjectFunc is a closure returning a test object (as JSON data structure), with the
+// focus value plugged in at the leaf.
+type SurroundingObjectFunc func(focus interface{}) (map[string]interface{}, FocusAccessorFunc, error)
+
+// NewRootObjectFunc returns a root closure of an object. The passed focus value
+// must be an object.
```

Comment:
is the passed object assumed to be acceptable to modify?

Reply:
If you mutate `surroundingObjectFunc(x)` (e.g. prune), `x` is also (potentially) mutated (e.g. pruned). That's the whole reason why we need this concept.

Reply:
that's fine, should just be called out

Reply:
done